forge-game/src/main/java/forge/game/card/Card.java

```diff
@@ -4386,6 +4385,52 @@ public final boolean hasIntensity() {
         return intensity > 0;
     }
 
+    private List<Pair<String, Map<String, Object>>> perpetual = new ArrayList<>();
+    public final boolean hasPerpetual() {
+        return !perpetual.isEmpty();
+    }
+    public final List<Pair<String, Map<String, Object>>> getPerpetual() {
+        return perpetual;
+    }
+
+    public final void addPerpetual(Pair<String, Map<String, Object>> p) {
+        perpetual.add(p);
+    }
+
+    public final void executePerpetual(Pair<String, Map<String, Object>> p) {
+        final String pType = p.getKey();
+        Map<String, Object> params = p.getValue();
+        if (pType.equals("NewPT")) {
+            addNewPT((Integer) params.get("Power"), (Integer) params.get("Toughness"), (long) 
+                params.get("Timestamp"), (long) 0);
+        } else if (pType.equals("PTBoost")) {
+            addPTBoost((Integer) params.get("Power"), (Integer) params.get("Toughness"), (long) 
+                params.get("Timestamp"), (long) 0);
+        } else if (pType.equals("Keywords")) {
+            addChangedCardKeywords((List<String>) params.get("AddKeywords"), Lists.newArrayList(), 
+                false, (long) params.get("Timestamp"), (long) 0);        
+        } else if (pType.equals("Types")) {
+            addChangedCardTypes((CardType) params.get("AddTypes"), (CardType) params.get("RemoveTypes"), 
+            false, (Set<RemoveType>) params.get("RemoveXTypes"), 
+            (long) params.get("Timestamp"), (long) 0, true, false);
+        } else if (pType.equals("Abilities")) {
+            addChangedCardTraits((CardTraitChanges) params.get("ChangedTraits"), 
+                (long) params.get("Timestamp"), (long) 0);
+        }
+    }
+
+    public final void setPerpetual(final Card oldCard) {
+        final List<Pair<String, Map<String, Object>>> perp = oldCard.getPerpetual();
+        perpetual = perp;
+        for (Pair <String, Map<String, Object>> p : perp) {
+            if (p.getKey().equals("Abilities")) {
```

Comment:
not sure why you added this here now while leaving the older part?

also you're still copying them with the old host, that's just wrong